varname/ignore.py

```diff
@@ -0,0 +1,343 @@
+                       frameinfo: Optional[inspect.FrameInfo] = None) -> None:
+    """Print the debug message for a given frame info object
+
+    Args:
+        msg: The debugging message
+        frameinfo: The FrameInfo object for the frame
+    """
+    from .utils import config
+    if not config.debug:
+        return
+    if frameinfo is not None:
+        msg = (f'{msg} [In {frameinfo.function!r} at '
+               f'{frameinfo.filename}:{frameinfo.lineno}]')
+    sys.stderr.write(f'[{__name__}] DEBUG: {msg}\n')
+
+class IgnoreElem(ABC):
+    """An element of the ignore list"""
+
+    def __init__(self, ignore: IgnoreElemType) -> None:
+        self.ignore = ignore
+
+    @abstractmethod
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        """Whether the frame matches the ignore element"""
+
+    @abstractmethod
+    def __repr__(self) -> str:
+        """Representation of the element"""
+
+class IgnoreModule(IgnoreElem):
+    """Ignore calls from a module or its submodules"""
+
+    def __init__(self, ignore: Union[ModuleType, str]) -> None:
+        super().__init__(ignore)
+        self.is_module = isinstance(ignore, ModuleType)
+        if self.is_module:
+            modfile = getattr(self.ignore, '__file__', None)
+            if not modfile or not path.isfile(modfile):
+                setattr(self.ignore, MODULE_IGNORE_ID_NAME,
+                        f'<varname-ignore-{id(self.ignore)}>')
+
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        frame = frameinfos[frame_no].frame
+
+        if self.is_module:
+            module = inspect.getmodule(frame)
+            if module:
+                return (module.__name__ == self.ignore.__name__ or
+                        module.__name__.startswith(f'{self.ignore.__name__}.'))
+
+            return (getattr(self.ignore, MODULE_IGNORE_ID_NAME, None) ==
+                    frame.f_globals.get(MODULE_IGNORE_ID_NAME, ''))
+
+        if path.isdir(self.ignore):
+            # ignore all calls from modules from a directory
+            # Used internally currently to ignore standard libraries
+            if not self.ignore.endswith('/'):
+                self.ignore = f'{self.ignore}/'
+            return frame.f_code.co_filename.startswith(self.ignore)
+
+        return frame.f_code.co_filename == self.ignore
+
+    def __repr__(self):
+        rep = self.ignore.__name__ if self.is_module else self.ignore
+        return f'<IgnoreModule({rep!r})>'
+
+class IgnoreFunction(IgnoreElem):
+    """Ignore a non-decorated function"""
+    def __init__(self, ignore: FunctionType) -> None:
+        super().__init__(ignore)
+        if (
+                # but this is annoying when the whole set (decorated function +
+                # var = func()) inside the same function
+                #'<locals>' in ignore.__qualname__ or # without functools.wraps
+                ignore.__name__ != ignore.__code__.co_name
```

Comment:
This works when they use `functools.wraps`, whereas `'<locals>' in ignore.__qualname__` would not work.

However if they don't use functools then a good check is `'<locals>' in Source.for_filename(ignore.__code__.co_filename).code_qualname(ignore.__code__)`

Reply:
Will this become annoying for the following:

```python
def sandbox():
  
  def func2():
    return varname()

  def func():
    return func2(ignore=func2)

  f = func()

sandbox()
```

If you use `'<locals>' in func2.__qualname__` to check if `func2` is decorated, but will be a false positive, since `func2.__qualname__ == 'sandbox.<locals>.func2'`.

Not sure if this helps: `'<locals>' in Source.for_filename(ignore.__code__.co_filename).code_qualname(ignore.__code__)`

Reply:
You're right, it would make a false positive for that.

Reply:
Then let's just leave this one here. In such a case, suppressing that warning is annoying.